Add fixtures

Garma00 opened the pull request:
- Add new package fixtures
- modify etl-func so it can run locally using azurite for message-content
- add fixtures to etl-func
- Add new types on io-messages-common package so they can be shared between projects

Resolves IOCOM-2247

Garma00 committed:
fix devcontainer command in readme

Garma00 committed:
Add etl fixtures

Garma00 committed:
add makeStorageAccountService function

Garma00 committed:
add config.ts inside blob-storage

Garma00 committed:
modify config to accept connection string in case of local development for storage account

Garma00 committed:
align local.settings.json

Garma00 committed:
update func.ts to use connection string in case of local development

Garma00 committed:
use valid markdown

Garma00 committed:
use connection string instead of credentials

Garma00 committed:
take the number of fixtures to generate as argument

Garma00 committed:
add readme

Garma00 committed:
add test for load fixture use case

Garma00 committed:
use test instead of it

Garma00 committed:
add tests for blob content loader

Garma00 committed:
add a test for the content load method

Garma00 committed:
add tests for metadata-loader

Garma00 committed:
remove .env.local.example in favour of .env.local

Garma00 committed:
add esling config and fix linter

Garma00 committed:
add .prettierignore

Garma00 committed:
fix code-review

lucamancosu commented on apps/etl-func/README.md:
````diff
@@ -68,3 +68,13 @@ Expected output:
+To insert fixtures simply run:
+
+```bash
+yarn fixtures:load <N>
````
Seams that running **yarn build** on the workspace folder is required before execute the **yarn fixtures:load** command. If i'm right i propose to add this instruction

Garma00 replied:
@lucamancosu running `yarn install` and `yarn build` in the root of the monorepo should be enough, I would avoid to add the instruction of building the single workspace just as we did with `io-messages-common`, do you agree?

lucamancosu replied:
Ok

lucamancosu commented on apps/etl-func/local.settings.json.example:
```diff
@@ -6,6 +6,7 @@
     "FUNCTIONS_WORKER_RUNTIME": "node",
     "MESSAGE_CONTENT_STORAGE_URI": "https://testgiovannist.blob.core.windows.net/",
     "MESSAGE_CONTENT_CONTAINER_NAME": "messages",
+    "MESSAGE_CONTENT_STORAGE_CONNECTION_STRING":"DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://io-messages-azurite-1:10000/devstoreaccount1;QueueEndpoint=http://localhost:10001/devstoreaccount1;TableEndpoint=http://localhost:10002/devstoreaccount1;",
```
probably we should replace io-messages-azurite-1  with localhost ?

Garma00 replied:
@lucamancosu `io-messages-azurite-1` is the name of the container, if we plan to make `devcontainer` as the standard way of testing the function locally I would leave it like that.
Anyway you made me notice that i left `localhost` in the last part of this connection string, if you agree with me I'll update it to use `io-messages-azurite-1`.

lucamancosu replied:
Using the devcontainer we can reference to each service with the relative compose service. In the case of io-messages-azurite-1 we should use **azurite**. 
The string connection becomes like this

"DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://azurite:10000/devstoreaccount1;QueueEndpoint=http://azurite:10001/devstoreaccount1;TableEndpoint=http://azurite:10002/devstoreaccount1;",
(tell me if it works)

But this connection string will works only in a docker network(as the one in the devcontainer environment) where the compose service name is resolved.

I prefer to avoid the name of the container(io-messages-azurite-1) because it can change in some circumstances. 

What do you think ?

lucamancosu commented on packages/etl-fixtures/README.md:
````diff
@@ -0,0 +1,13 @@
+running the following command inside the `etl-func` app:
+
+```bash
+yarn fixtures:load
````
Seams that running yarn build on the workspace folder is required before execute the yarn fixtures:load command. If i'm right i propose to add this instruction

Garma00 replied:
https://github.com/pagopa/io-messages/pull/184#discussion_r2014080106

lucamancosu commented on apps/etl-func/README.md:
````diff
@@ -68,3 +68,13 @@ Expected output:
 ```bash
 it works!
 ```
+
````
there is probably a typo in the file

cp local.settings.json.example .local.settings.json

should be

**cp local.settings.json.example local.settings.json**

Garma00 replied:
@lucamancosu nice catch, I fixed it [e095708](https://github.com/pagopa/io-messages/pull/184/commits/e0957087619a5bf3223bd31448e24b12f78f2f61)

lucamancosu commented:
Should we create a script for the creation of the table for the tracking of the errors in the account storage ?

Garma00 commented:
> Should we create a script for the creation of the table for the tracking of the errors in the account storage ?

@lucamancosu nice point! I guess we can add a `createIfNotExist` as we did for the blob container in the index file, I'll implement this, thanks!

Garma00 committed:
fix readme local.settings.json

Garma00 committed:
add create table method in index.ts

Garma00 committed:
remove .env.local

Now enviroment variables are directly taken from the local.settings.json
in the etl-func app

Garma00 committed:
update readme

Garma00 committed:
fix linter

Garma00 committed:
Merge branch 'main' into feats/add-etl-fixtures

Garma00 committed:
add return in load method so the caller will wait till the Promise.all completition

Garma00 commented on packages/etl-fixtures/src/adapters/cosmos/metadata-loader.ts:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  async load(metadata: MessageMetadata[]): Promise<void> {
+    Promise.all(metadata.map((m) => this.container.items.create(m)));
```
[Here](https://github.com/pagopa/io-messages/pull/184/commits/188f4a6c7fe6a0c06e036c3e8ecb7f25770c5e37) I now return the Promise.all result so the caller can wait for the completion using await.
I don't want to use await here because I believe it should be up to the caller to decide whether to wait or not.

lucamancosu commented on apps/etl-func/README.md:
````diff
@@ -68,3 +68,13 @@ Expected output:
+To insert fixtures simply run:
+
+```bash
+yarn fixtures:load <N>
````
It seams i cant start the fixture process because i'm missing FIXTURES_INCLUDE_PAYMENTS and FIXTURES_INCLUDE_REMOTE_CONTENT. 
I think this variables are not mandatory and they must have a default value. 
If there is a reason to make them mandatory then i suggest to add them to the local.settings.example

Garma00 replied:
@lucamancosu those values should not be mandatory, they already have a default value take a look [here](https://github.com/pagopa/io-messages/blob/feats/add-etl-fixtures/packages/etl-fixtures/src/adapters/config.ts#L7)

lucamancosu replied:
Still can't start the fixtures script without declaring FIXTURES_INCLUDE_PAYMENTS and FIXTURES_INCLUDE_REMOTE_CONTENT on local.settings

Garma00 committed:
generalize fixtures so those can be used for any app

Garma00 committed:
add example links to README

Garma00 committed:
Merge branch 'main' into feats/add-etl-fixtures

lucamancosu commented on packages/fixtures/src/adapters/cosmos/metadata-loader.ts:
```diff
@@ -0,0 +1,74 @@
+  async load(
+    metadata: MessageMetadata[],
+  ): Promise<ItemResponse<MessageMetadata>[]> {
+    const r = Promise.all(metadata.map((m) => this.container.items.create(m)));
```
I think we must await there.

Garma00 replied:
I think it should be up to the caller weather to wait fot this load operation or not, we are here returning the Promise.all result so the caller can wait for it, in fact in the use case we have an `await`

Garma00 committed:
return Promise.all result directly instead of using a const

lucacavallaro commented on apps/etl-func/package.json:
```diff
@@ -9,6 +9,7 @@
   "scripts": {
     "build": "esbuild --bundle ./src/func.ts --format=esm --platform=node --target=node20 --outdir=dist --packages=external",
     "build:watch": "yarn build --watch",
+    "fixtures": "tsx scripts/fixtures.ts",
```
we can just use `js` and avoid `tsx`, what you think?

Garma00 replied:
Well, good idea, did it here [3532cd4](https://github.com/pagopa/io-messages/pull/184/commits/3532cd46a662a481556d1f82043b722b06388bd5)

lucacavallaro commented on apps/etl-func/scripts/fixtures.ts:
```diff
@@ -0,0 +1,42 @@
+  });
+} else {
+  throw new Error("ETL function not found");
+}
```
why are you loading these variables onto `environment` when on line 30 you use them as variables? you can skip this part

Garma00 replied:
@lucacavallaro Those variables are not accessible from process.env unless we load them from local.settings.json. Is there a cleaner way to do this?

lucacavallaro replied:
no, what I meant is that maybe we don't need to load these variable into environment

you can simply parse them to `settings` const and then pass to `loadFixturesOptionsSchema.parse` as `settings.MY_VAR`

Garma00 replied:
Ok sorry now I understand, please take a look here [7fbba1e](https://github.com/pagopa/io-messages/pull/184/commits/7fbba1e3c9637bc654dd62ebfe3c3f20fe18978e)

lucacavallaro commented on packages/fixtures/src/adapters/blob/content-loader.ts:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  async load(messageContents: MessageContentWithIds[]): Promise<void> {
+    Promise.all(
```
you are returning the wrong promise here :)

Garma00 replied:
@lucacavallaro You're right — I was reviewing this method again and I noticed a few things:
1. We're returning void here
2. In the load method for metadata, we're returning a type imported from the Cosmos SDK

I think a cleaner way to write this method would be to define a custom type like:
```ts
{ success: boolean; error?: Error }
```
This way, we can establish a clear 'port' between the use case and the adapter. Take a look at [05c80a5](https://github.com/pagopa/io-messages/pull/184/commits/05c80a53d92b226e2280add1349d3092dd771edf), do you agree?

lucacavallaro replied:
how is this interface different than returning `void` if OK and throwing `Error` if not?

Garma00 replied:
Ok I agree, I simplified the code [ab56815](https://github.com/pagopa/io-messages/pull/184/commits/ab5681588687a2e8aa3f5c0aa414eefb5cbc0c0d)

lucacavallaro commented on packages/fixtures/src/adapters/config.ts:
```diff
@@ -0,0 +1,21 @@
+  cosmosEndpoint: z.string().url(),
+  cosmosMessageContainerName: z.string().min(1),
+  includePayments: booleanFromString,
+  includeRemoteContents: booleanFromString,
```
why `strings` and not `boolean`?

Garma00 replied:
From process.env we have to parse the string "true" into a boolean so I creted this utility

Garma00 replied:
@lucacavallaro since I made the changes you suggested [here](https://github.com/pagopa/io-messages/pull/184#discussion_r2042392557) now we can simply use boolean, take a look here [43cb5ee](https://github.com/pagopa/io-messages/pull/184/commits/43cb5eea71c37eff060f33f2566c25e9cc3c4a1e)

lucacavallaro commented on packages/io-messages-common/src/types/log.ts:
```diff
@@ -0,0 +1,5 @@
+  error: (message: string) => void;
+  info: (message: string) => void;
+  log: (message: string) => void;
+}
```
is this abstraction really needed?

Garma00 replied:
We can omit this and use the pino Logger as we did [here](https://github.com/pagopa/io-messages/blob/main/apps/etl-func/src/adapters/message.ts#L26C13-L26C19) for example, on the other hand I think that this approach goes against the port and adapter principle, what do you thinK @lucacavallaro?

lucacavallaro replied:
this is a little utility, we can relax this kind of constraints and avoid unnecessary abstractions

Garma00 replied:
[032e652](https://github.com/pagopa/io-messages/pull/184/commits/032e652d666e877949346d834ed900def1001ca0)

Garma00 committed:
add a port between use case and adapters

Garma00 committed:
replace ts with js in the etl script to avoid tsx

Garma00 committed:
Merge branch 'main' into feats/add-etl-fixtures

Garma00 committed:
directly gett settings instead of using process.env

Garma00 committed:
remove booleanFromString

Garma00 committed:
remove custom Logger type

Garma00 committed:
return void using await in load functions

lucacavallaro commented on packages/fixtures/src/adapters/blob/content-loader.ts:
```diff
@@ -48,27 +46,15 @@ export class BlobContentLoader implements ContentLoader {
-      };
-    }
+  async load(messageContents: MessageContentWithIds[]): Promise<void> {
+    await Promise.all(
```
```suggestion
    return Promise.all(
```

Garma00 replied:
Sorry I don't get the point, I agree that it should be better to return the promise so the caller can decide weather to wait or not but [here](https://github.com/pagopa/io-messages/pull/184#discussion_r2042394850) we said to return void

lucacavallaro commented on packages/fixtures/src/adapters/cosmos/metadata-loader.ts:
```diff
@@ -65,19 +63,7 @@ export class CosmosMetadataLoader implements MetadataLoader {
-      };
-    }
+  async load(metadata: MessageMetadata[]): Promise<void> {
+    await Promise.all(metadata.map((m) => this.container.items.create(m)));
```
```suggestion
    return Promise.all(metadata.map((m) => this.container.items.create(m)));
```

lucacavallaro commented on apps/etl-func/scripts/fixtures.js:
```diff
@@ -18,25 +18,24 @@ const etlFunctionPath = resolve(__dirname, "../local.settings.json");
-}
+const getSettings = (settingsPath) => {
+  if (existsSync(settingsPath)) {
+    return JSON.parse(readFileSync(etlFunctionPath, "utf8")).Values;
```
```suggestion
    return JSON.parse(readFileSync(settingsPath, "utf8")).Values;
```

lucacavallaro commented on apps/etl-func/scripts/fixtures.js:
```diff
@@ -18,25 +18,24 @@ const etlFunctionPath = resolve(__dirname, "../local.settings.json");
+  includeRemoteContents: settings.FIXTURES_INCLUDE_REMOTE_CONTENT,
 });
 
 loadFixtures(fixturesToGenerate, { ...options, aadCredentials });
```
why we don't `await` here?

Garma00 committed:
Merge branch 'main' into feats/add-etl-fixtures

Garma00 committed:
refactor for clean architecture

Garma00 committed:
add unit test for message and message-generator adapters

Garma00 committed:
use the path passed as parameter in getSettings

Garma00 committed:
wait for loadFixtures to complete

Garma00 commented:
@lucacavallaro I made a refactor in order to respect clean architecture principles as we discussed

Garma00 committed:
fix linter

Garma00 committed:
fix linter in io-messages-common

Garma00 committed:
fix code-review script

Garma00 committed:
import types from io-messages-common in etl func

Garma00 committed:
Merge branch 'main' into feats/add-etl-fixtures

Garma00 commented:
This feature is no longer needed.
